README
NAME
    PkgConfig::LibPkgConf - Interface to .pc file interface via libpkgconf

SYNOPSIS
     use PkgConfig::LibPkgConf;
 
     if(pkgconf_exists('libarchive'))
     {
       my $version = pkgconf_version('libarchive');
       my $cflags  = pkgconf_cflags('libarchive');
       my $libs    = pkgconf_libs('libarchive');
     }

DESCRIPTION
    Many libraries in compiled languages such as C or C++ provide ".pc"
    files to specify the flags required for compiling and linking against
    those libraries. Traditionally, the command line program "pkg-config" is
    used to query these files. This module provides a Perl level API using
    "libpkgconf" to these files.

    This module provides a simplified interface for getting the existence,
    version, cflags and library flags needed for compiling against a
    package, using the default compiled in configuration of "pkgconf". For a
    more powerful, but complicated interface see
    PkgConfig::LibPkgConf::Client. In addition, PkgConfig::LibPkgConf::Util
    provides some useful utility functions that are also provided by
    "pkgconf".

FUNCTIONS
  pkgconf_exists
     my $bool = pkgconf_exists $package_name;

    Returns true if the package is available.

    Exported by default.

  pkgconf_version
     my $version = pkgconf_version $package_name;

    Returns the version of the package, if it exists. Will throw an
    exception if not found.

    Exported by default.

  pkgconf_cflags
     my $cflags = pkgconf_cflags $package_name;

    Returns the compiler flags for the package, if it exists. Will throw an
    exception if not found.

    Exported by default.

  pkgconf_cflags_static
     my $cflags = pkgconf_cflags_static $package_name;

    Returns the static compiler flags for the package, if it exists. Will
    throw an exception if not found.

  pkgconf_libs
     my $libs = pkgconf_libs $package_name;

    Returns the linker library flags for the package, if it exists. Will
    throw an exception if not found.

    Exported by default.

  pkgconf_libs_static
     my $libs = pkgconf_libs_static $package_name;

    Returns the static linker library flags for the package, if it exists.
    Will throw an exception if not found.

SUPPORT
    IRC #native on irc.perl.org

    Project GitHub tracker:

    <https://github.com/plicease/PkgConfig-LibPkgConf/issues>

    If you want to contribute, please open a pull request on GitHub:

    <https://github.com/plicease/PkgConfig-LibPkgConf/pulls>

SEE ALSO
    The best entry point to the low level "pkgconf" interface can be found
    via PkgConfig::LibPkgConf::Client.

    Alternatives include:

    PkgConfig
        Pure Perl implementation of "pkg-config" which can be used from the
        command line, or as an API from Perl. Does not require pkg-config in
        your path, so is a safe dependency for CPAN modules.

    ExtUtils::PkgConfig
        Wrapper for the "pkg-config" command line interface. This module
        will fail to install if "pkg-config" cannot be found in the "PATH",
        so it is not safe to use a dependency if you want your CPAN module
        to work on platforms where "pkg-config" is not installed.

    Alien::Base
        Provides tools for building non-Perl libraries and making them
        dependencies for your CPAN modules, even on platforms where the
        non-Perl libraries aren't already installed. Includes hooks for
        probing "pkg-config" ".pc" files using either "pkg-config" or
        PkgConfig.

ACKNOWLEDGMENTS
    Thanks to the "pkgconf" developers for their efforts:

    <https://github.com/pkgconf/pkgconf/graphs/contributors>

AUTHOR
    Graham Ollis

    Contributors:

    A. Wilcox (awilfox)

COPYRIGHT AND LICENSE
    This software is copyright (c) 2016 Graham Ollis.

    This is free software; you may redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for PkgConfig-LibPkgConf

  - Work around for bug with pkgconf_compare_version
    on arm64 (gh#9)

0.10 2018-12-26 21:29:54 -0500
  - Explicitly require EUMM 6.98
  - Explicitly require Test::More 0.98

0.09 2018-06-25 06:55:46 -0400
  - Compatability with pkgconf 1.5.0
  - Require pkgconf 1.5.0 (via Alien::pkgconf 0.12 if necessary)

0.08 2017-12-27 09:38:07 -0500
  - Compatability with pkgconf 1.3.90
  - Require pkgconf 1.3.90

0.07 2017-03-15 10:45:03 -0400
  - Use correct CCFLAGS for Perl ABI (ppisar++ gh#5)

0.06 2017-03-02 17:43:40 -0500
  - Update for 1.3.0 API

0.05 2017-01-31 08:25:00 -0500
  - Expose pkgconf_path_relocate as PkgConfig::LibPkgConf::Util::path_relocate

0.04 2017-01-22 02:43:49 -0500
  - Update XS to work with libpkgconf 1.2.0
  - Requires libpkgconf 1.2.0
  - No interface change

0.03 2017-01-18 08:32:50 -0500
  - Fixed linker issues on FreeBSD

0.02 2017-01-03 11:24:13 -0500
  - Added pkgconf_cflags_static, pkgconf_libs_static functions
    to PkgConfig::LibPkgConf

0.01 2016-12-31 16:45:06 -0500
  - Initial version


Makefile.PL
use strict;
use warnings;

BEGIN {
  unless(eval q{ use 5.008001; 1 })
  {
    print "Sorry!  Perl 5.8.1 or better required\n";
    exit;
  }
}

use ExtUtils::MakeMaker 6.98;
use Config;

my $cc = $Config{cc};
my $ld = $Config{ld};
my $alien = 0;
my $alien_ver   = '0.12';
my $pkgconf_min = '1.5.0';

sub try
{
  my($name) = @_;

  if(eval q{ use Alien::pkgconf $alien_ver; 1 })
  {
    print "Probe with Alien::pkgconf\n";
    return (Alien::pkgconf->cflags, Alien::pkgconf->libs, Alien::pkgconf->version);
  }

  foreach my $cmd ($ENV{PKG_CONFIG}, qw( pkgconf pkg-config ppkg-config pkg-config.pl ))
  {
    next unless defined $cmd;
    do {
      no warnings 'exec';
      system $cmd, '--exists', $name;
    };
    if($? == 0)
    {
      my @l = eval {
        system $cmd, "--atleast-version=$pkgconf_min", 'libpkgconf';
        die if $?;
        my $cflags = `$cmd --cflags $name`;
        die if $?;
        my $libs   = `$cmd --libs   $name`;
        die if $?;
        my $version   = `$cmd --modversion   $name`;
        die if $?;
        chomp $version;
        print "Probe with $cmd\n";
        chomp $cflags;
        chomp $libs;
        ($cflags, $libs, $version);
      };
      return @l unless $@;
    }
  }

  if(eval { require PkgConfig; PkgConfig::Version->can('new') })
  {
    my $package = PkgConfig->find($name);
    if(!$package->errmsg)
    {
      my $version = PkgConfig::Version->new($package->pkg_version);
      my $need    = PkgConfig::Version->new($pkgconf_min);
      if($version >= $need)
      {
        print "Probe with PkgConfig.pm\n";
        my $cflags = $package->get_cflags . ' ';
        my $libs   = $package->get_ldflags . ' ';
        return ($cflags, $libs, $version);
      }
    }
  }

  if(eval { require ExtUtils::PkgConfig; 1 })
  {
    # ick.  EU::PC cannot be required unconditionally
    # because it doesn't install without pkg-config.
    # also it is very unecessarily noisy!
    if(ExtUtils::PkgConfig->atleast_version($name,$pkgconf_min))
    {
      my %info = eval { ExtUtils::PkgConfig->find($name) };
      if(!$@)
      {
        print "Probe with ExtUtils::PkgConfig\n";
        return ($info{cflags}, $info{libs}, $info{modversion});
      }
    }
  }

  $alien = 1;
  $cc = '$(FULLPERL) script/cc_wrapper.pl';
  $ld = '$(FULLPERL) script/ld_wrapper.pl';
  return ('','', '');
}

my($ccflags, $ldflags, $version) = try 'libpkgconf';

print "($ccflags, $ldflags, $version)\n";

$ccflags = $ccflags ? "$ccflags $Config{ccflags}" : $Config{ccflags};
$ccflags .= " -DMY_PKGCONF_VERSION=$version" unless $alien;

my $lddlflags = $Config{lddlflags};
my $optimize  = $Config{optimize};

# This is not Platypus, but lets honor the debug env
if($ENV{FFI_PLATYPUS_DEBUG})
{
  print "-lddlflags=$lddlflags\n";
  $lddlflags =~ s/-O[0-9]?/-g3/g;
  print "+lddlflags=$lddlflags\n";

  print "-optimize=$optimize\n";
  $optimize =~ s/-O[0-9]?/-g3/g;
  print "+optimize=$optimize\n";
}

if($^O eq 'freebsd' || $^O eq 'dragonfly')
{
  if($lddlflags =~ m{-L/usr/local/lib})
  {
    require Text::ParseWords;
    my %seen;
    foreach my $flag (Text::ParseWords::shellwords($ldflags))
    {
      next if $seen{$flag};
      $seen{$flag}++;
      next unless $flag =~ /^-L/;

      print "-lddlflags=$lddlflags\n";
      $lddlflags =~ s{(-L/usr/local/lib)}{$flag $1}g;
      print "+lddlflags=$lddlflags\n";

    }
  }
}

my %args = (
  ABSTRACT           => 'Perl bindings for libpkgconf',
  AUTHOR             => [ 'Graham Ollis <plicease@cpan.org>' ],
  DISTNAME           => 'PkgConfig-LibPkgConf',
  LICENSE            => 'perl',
  MIN_PERL_VERSION   => '5.008001',
  NAME               => 'PkgConfig::LibPkgConf',
  VERSION_FROM       => 'lib/PkgConfig/LibPkgConf.pm',

  CC                 => $cc,
  LD                 => $ld,

  CCFLAGS            => $ccflags,
  LIBS               => $ldflags,
  LDDLFLAGS          => $lddlflags,
  OPTIMIZE           => $optimize,

  CONFIGURE_REQUIRES => {
    "ExtUtils::MakeMaker" => '6.98',
  },
  BUILD_REQUIRES     => {
    ($alien ? ('Alien::pkgconf' => $alien_ver) : ()),
  },
  TEST_REQUIRES      => {
    'Test::More' => '0.94',
  },
  PREREQ_PM          => {
    #
  },

  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      repository => {
        type => 'git',
        url  => "http://github.com/plicease/PkgConfig-LibPkgConf.git",
        web  => "http://github.com/plicease/PkgConfig-LibPkgConf",
      },
      bugtracker => {
        web => "http://github.com/plicease/PkgConfig-LibPkgConf/issues",
      },
      homepage => "https://metacpan.org/pod/PkgConfig-LibPkgConf",
      x_MailingList => "https://groups.google.com/forum/#!forum/perl5-alien",
      x_IRC => "irc://irc.perl.org/#native",
    },
    dynamic_config => 1,
  },

  clean => {
    FILES => '*.log *.bak',
  }
);

WriteMakefile(%args);
